onnxruntime/core/optimizer/unsqueeze_elimination.cc

```diff
@@ -2,104 +2,108 @@
-      continue;
-    }
+  std::vector<int64_t> axes;
+  for (int i = 0; i < attr->ints_size(); i++) {
+    axes.push_back(static_cast<int64_t>(attr->ints(i)));
+  }
 
-    const onnxruntime::NodeAttributes& attributes = node.GetAttributes();
-    const ONNX_NAMESPACE::AttributeProto* attr = &attributes.find("axes")->second;
-    if (attr == nullptr || attr->type() != AttributeProto_AttributeType_INTS) {
-      continue;
-    }
+  // Generate new dims
+  NodeArg* input_def = node.MutableInputDefs()[0];
+  const ONNX_NAMESPACE::TensorProto* tensor_proto = nullptr;
+  graph.GetInitializedTensor(input_def->Name(), tensor_proto);
+  if (tensor_proto == nullptr) {
+    return Status::OK();
+  }
+  std::vector<int64_t> new_dims(axes.size() + tensor_proto->dims().size(), 0);
+  if (new_dims.size() >= std::numeric_limits<int>::max()) {
+    return Status(ONNXRUNTIME, FAIL, "index out of range");
+  }
 
-    // Get attribute of "axes"
-    std::vector<int64_t> axes;
-    for (int i = 0; i < attr->ints_size(); i++) {
-      axes.push_back(static_cast<int64_t>(attr->ints(i)));
-    }
+  for (int64_t axis : axes) {
+    new_dims[axis] = 1;
+  }
 
-    // Generate new dims
-    NodeArg* input_def = node.MutableInputDefs()[0];
-    const ONNX_NAMESPACE::TensorProto* tensor_proto = nullptr;
-    graph.GetInitializedTensor(input_def->Name(), tensor_proto);
-    if (tensor_proto == nullptr) {
-      continue;
+  auto begin = tensor_proto->dims().cbegin();
+  for (auto& axis : new_dims) {
+    if (axis == 0) {
+      axis = *begin++;
     }
-    std::vector<int64_t> new_dims(axes.size() + tensor_proto->dims().size(), 0);
-    if (new_dims.size() >= std::numeric_limits<int>::max())
-      return Status(ONNXRUNTIME, FAIL, "index out of range");
+  }
 
-    for (int64_t axis : axes) {
-      new_dims[axis] = 1;
-    }
+  // Update shape of tensor proto
+  ONNX_NAMESPACE::TensorProto new_tensor_proto(*tensor_proto);
 
-    auto begin = tensor_proto->dims().cbegin();
-    for (auto& axis : new_dims) {
-      if (axis == 0) {
-        axis = *begin++;
-      }
+  for (int i = 0; i < static_cast<int>(new_dims.size()); i++) {
+    if (i < tensor_proto->dims().size()) {
+      new_tensor_proto.set_dims(i, new_dims[i]);
+    } else {
+      new_tensor_proto.add_dims(new_dims[i]);
     }
+  }
+  graph.RemoveInitializedTensor(input_def->Name());
+  graph.AddInitializedTensor(new_tensor_proto);
 
-    // Update shape of tensor proto
-    ONNX_NAMESPACE::TensorProto new_tensor_proto(*tensor_proto);
+  // Update shape of NodeArg
+  TensorShapeProto shape;
+  for (auto dim : new_dims) {
+    shape.add_dim()->set_dim_value(dim);
+  }
+  input_def->SetShape(shape);
+
+  // Keep all output nodes in a list (after removing all edges below, there will be no other way to access them; 
+  // I cannot first update the output defs, because edge removal will not work afterwards).
+  std::vector<NodeIndex*> output_nodes_idx;
```

Comment:
>output_nodes_idx [](start = 26, length = 16)

can you use graph_utils::RemoveSingleInSingleOutNode ?

Reply:
Unfortunately, I cannot.
Unsqueeze here has an initializer as input, which means that it does not have an edge.
Also, Unsqueeze has multiple outputs, while this method deals with single output.
Moreover, there is the complication I mention in the comment (RemoveEdge does not work if you have previously updated the NodeArg of one of the two endpoints of the edge). I will most probably remove that check from the RemoveEdge in the next commit.
That said I was thinking to update the RemoveSingleInSingleOut method to accept multiple outputs. Will see if I do it here or in another PR.

Reply:
I think RemoveSingleInSingleOut is overly restrictive as it shouldn't matter how many other nodes consume the output. Is it possible to make a more generic 'remove node' method based on what's in RemoveSingleInSingleOut currently, that just repeats the update logic for each input and output?  Ideally we have a common piece of node removal logic (which should really be provided by the Graph class) so that we ensure subgraph checking/update happens. 

---
In reply to: [267893050](https://github.com/Microsoft/onnxruntime/pull/670#discussion_r267893050) [](ancestors = 267893050)

Reply:
And we need to ensure that the subgraph check to happens for any potential node removal one way or another.

---
In reply to: [268067765](https://github.com/Microsoft/onnxruntime/pull/670#discussion_r268067765) [](ancestors = 268067765,267893050)

Reply:
I agree. I had in mind to do this (either in this PR or in another).
For sure single-in-multiple-out is possible.
I will see if I can do something like this when the input is an initializer too (like in the UnsqueezeElimination case). Or at least reuse some of this logic. I need to remove the check in the RemoveEdge and then probably it will get cleaner.

Reply:
I will open a follow-up PR to make this more generic.